Split all length prefix framing into own package

drawks opened the pull request:
Doing length prefixed messaging framing always seemed like something best left up to an isolated utility package to me. So I factored it out into go-carbon/helper/framing.

Commit:
Factor length prefixed network ops to own package

- Include helper/framing package derived from github.com/eproxus/framing
- updated carbonlink to use new framing package
- updated carbonlink error response to use framed connection
- updated receiver to use new framing package for tcp+pickle
- updated tests to reflect updates

drawks commented:
Failed check is unrelated to this change...

drawks commented:
Can I get a merge or feedback on this PR?

redbaron commented on helper/framing/conn.go:
```diff
@@ -0,0 +1,168 @@
+			return total, err
+		}
+		total += n
+		b = b[size:]
```
does `Conn.Read` guarantee to return `n == size` when `err == nil`?

redbaron commented on helper/framing/conn.go:
```diff
@@ -0,0 +1,168 @@
+		size := min(f.bytesLeft, len(b))
+
+		n, err := f.Conn.Read(b[:size])
+		if err != nil {
```
is `n == 0` when `err != nil` in all the cases?

redbaron commented on helper/framing/conn.go:
```diff
@@ -0,0 +1,168 @@
+		f.bytesLeft = frame_size
+	}
+
+	for len(b) > 0 && f.bytesLeft > 0 {
```
Can be calculated once outside the loop:

``` go
bytesToRead := min(f.bytesLeft, len(b))
for bytesToRead > 0 {
    ...
    bytesToRead -= n
}
```

redbaron replied:
Will `io.ReadFull` work here?

```
return io.ReadFull(f.Conn, b[:min(f.bytesLeft, len(b)])
```

lomik commented on helper/framing/conn_test.go:
```diff
@@ -0,0 +1,292 @@
+		defer framed.Close()
+
+		for i := 1; i <= 2; i++ {
+			if _, err := fmt.Fprintf(framed, message); err != nil {
```
Why Fprintf? `message` is not "format"

lomik commented on helper/framing/conn_test.go:
```diff
@@ -0,0 +1,292 @@
+	port := l.Addr().(*net.TCPAddr).Port
+
+	go func() {
+		conn, err := net.Dial("tcp", ":"+strconv.Itoa(port))
```
May be

```
conn, err := net.Dial("tcp", l.Addr().String())
```

lomik commented on helper/framing/conn_test.go:
```diff
@@ -0,0 +1,292 @@
+	conn := &fake{ReadWriter: bytes.NewBuffer([]byte{4, 0, 1, 2, 3})}
+	framed, _ := framing.NewConn(&fake{ReadWriter: conn}, 1, nil)
+
+	first_b := make([]byte, 6)
```
"first_b", "first_n", "first_err", "third_n", "third_err", "framed_conn" and many other variable names are bad. https://golang.org/doc/effective_go.html#mixed-caps

drawks commented:
lots of good notes here. I'll take a crack at them sometime this week.

Commit:
Integrated updates per PR notes

- renamed vars with underbars
- simplified code for framing.Conn.Read to use io.ReadFull
- ensured a clean lint pass for helper/framing

drawks commented:
I /think/ I addressed all the commented code. More comments and/or merge?

drawks commented:
nev mind, gotta rebase. Will take a look tonight

drawks commented:
alright, got everything rebased and squashed my commits in a sensible way. c-i failure appears to be unrelated to this pr (the long standing race in cache testing and also the throttle test instability)

lomik commented on helper/framing/conn.go:
```diff
@@ -0,0 +1,157 @@
+	total := 0
+
+	if f.bytesLeft <= 0 {
+		frame_size, err := f.readSize()
```
Please, fix "frame_size" variable name. Try to run golint

drawks commented:
Updated branch to resolve outstanding style issues. Please consider merge.

drawks commented:
@lomik ping!
